### PR TITLE
Fix loading grayscale emissive map 

### LINKED ISSUE #499: Emissive map no longer works in ign-rendering6 (ogre2)
## Environment
* OS Version: Ubuntu 18.04
* Source or binary build?
  source build: `ign-rendering6` branch
 
## Description
* Expected behavior: objects are rendered with emissive map
* Actual behavior: emissive map has no effect

I think the feature is broken when we ported from ogre 2.1 to 2.2.

## Steps to reproduce
1. Test with ignition gazebo's fuel_textured_mesh.sdf world
  ```
  ign gazebo -v 4 fuel_textured_mesh.sdf
  ```
2. The tubes in the `Tube Light` model do not appear to be lit when emissive map is used.

Here is a comparison of the Tube Light model in Ignition Fortress (top) and Dome (bottom)

emissive_map_fortress_vs_dome

## Things I've tried

The call to set ogre2 emissive map is done here:
https://github.com/ignitionrobotics/ign-rendering/blob/ign-rendering6/ogre2/src/Ogre2Material.cc#L473

* I verified that `this->ogreDatablock->getUseEmissiveAsLightmap()` returns `false` and `this->ogreDatablock->_hasEmissive()` returns `true`
* setting emissive values (`this->ogreDatablock->setEmissive`) works 
* I tried debugging the code in [800.PixelShader_piece_ps.any](https://github.com/ignitionrobotics/ign-rendering/blob/ign-rendering6/ogre2/src/media/Hlms/Pbs/Any/Main/800.PixelShader_piece_ps.any#L507-L526) for applying emissive map and it looks like the `emissive_map` property is not set hence it does not affect the final color.

### PR DESCRIPTION
**Pull request by iche033**
# 🦟 Bug fix

Fixes #499 

## Summary

Continuing the effort from #501, this PR fixes loading grayscale emissive map by converting it to RGB color texture so that the visual with the emissive map is not incorrectly rendered as red color.

Test by lauching ign gazebo with the `fuel_textured_mesh.sdf` world file.

Before:

tube_grayscale_emissive_before

After:

tube_grayscale_emissive_after

## Alternative considered

The other option considered is to some customization in the Pbs Hlms pixel shader and just apply only the r channel of the emissive texture to the final color, .e.g. [changing the code here](https://github.com/ignitionrobotics/ign-rendering/blob/ign-rendering6/ogre2/src/media/Hlms/Pbs/Any/Main/800.PixelShader_piece_ps.any#L521) to `finalColour += emissiveCol.rrr;`. 

I went with the approach in this PR mainly to keep the changes within one function without touching the shaders so we don't deviate too much from the original Pbs shader code we took from ogre. This approach was also easier for me to to implement.

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

File: ogre2/src/Ogre2Material.cc [hunk 1]
```diff
@@ -636,16 +637,82 @@ void Ogre2Material::SetTextureMapImpl(const std::string &_texture,
...
+      std::string p = common::parentPath(_texture);
+      // set a custom name for the rgb texture by appending ign_ prefix
+      std::string rgbTexName = "ign_" + baseName;
+      baseName = rgbTexName;
+      std::string filename = common::joinPaths(p, rgbTexName);
+      auto tex = textureMgr->findTextureNoThrow(rgbTexName);
+      if (!tex)
+      {
+        ignmsg << "Grayscale emissive texture detected. Converting to RGB: "
+            << rgbTexName << std::endl;
```
Comment (ahcorde) [inline ogre2/src/Ogre2Material.cc:662]:
```suggestion
        ignmsg << "Grayscale emissive texture detected. Converting to RGB: "
               << rgbTexName << std::endl;
```

File: ogre2/src/Ogre2Material.cc [hunk 2]
```diff
@@ -636,16 +637,82 @@ void Ogre2Material::SetTextureMapImpl(const std::string &_texture,
...
+  // workaround for grayscale emissive texture
+  // convert to RGB otherwise the emissive map is rendered red
+  if (_type == Ogre::PBSM_EMISSIVE &&
+      !this->ogreDatablock->getUseEmissiveAsLightmap())
+  {
+    common::Image img(_texture);
+    // check for 8 bit pixel
+    if (img.BPP() == 8u)
+    {
+      std::string p = common::parentPath(_texture);
```
Comment (ahcorde) [inline ogre2/src/Ogre2Material.cc:653]:
```suggestion
      std::string parentDirectory = common::parentPath(_texture);
```
more descriptive name?

**Review by ahcorde — CHANGES_REQUESTED** [review]

**Comment by darksylinc** [comment]
Sounds like this is something that should be fixed from Upstream.

Detect if emissive is greyscale, then use a shader with the right swizzle

File: ogre2/src/Ogre2Material.cc (hunk 2 above)
Comment (iche033) [reply ogre2/src/Ogre2Material.cc]: a6796ac

File: ogre2/src/Ogre2Material.cc (hunk 1 above)
Comment (iche033) [reply ogre2/src/Ogre2Material.cc]: a6796ac

**Comment by iche033** [comment]
great, thanks for implementing grayscale emissive map support upstream! I'm going to try get this in first, and remove later when I get time to port the fix over to our fork (and package / release new debs etc)

**Review by ahcorde — APPROVED** [review]

**Pull request merged**